### PR TITLE
[BugFix] fix bugs from reusing common expressions with nested high-order functions

### PR DESCRIPTION
**Pull request by fzhedu**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->

1. introduce `isLambdaDependent` to control whether reuse lambda-dependent expressions.

        // isLambdaDependent means whether an expression is dependent on outer lambda arguments,
        // it works when not reuse lambda arguments. For example, when reuseLambda is false, for expression
        // array_map(x -> x+1, array), the x+1 is lambda dependent, but array_map(x -> x+1, array) is not. 

2. introduce `hasLambdaFunction` to eable specific logic for lambda functions, reducing extra costs for non-lambda expressions in most cases.

3. add `needReuseLambdaExpr` for projection to avoid rewrite if the projection without lambda functions.


TODO: found a problem that two identical expressions, their sub expressions also listed in common expressions, a example listed below,  `<slot 7>` and `<slot 8>` should remove, only keep `<slot 9>`. This problem can be alleviated by reusing alias expressions in the future work.



```
mysql> explain select aa+1, ceil(aa) from (select abs(a+pv) as aa from t)A;
+-----------------------------------------------------------------------------+
| Explain String                                                              |
+-----------------------------------------------------------------------------+
| PLAN FRAGMENT 0                                                             |
|  OUTPUT EXPRS:5: expr | 6: ceil                                             |
|   PARTITION: UNPARTITIONED                                                  |
|                                                                             |
|   RESULT SINK                                                               |
|                                                                             |
|   2:EXCHANGE                                                                |
|                                                                             |
| PLAN FRAGMENT 1                                                             |
|  OUTPUT EXPRS:                                                              |
|   PARTITION: RANDOM                                                         |
|                                                                             |
|   STREAM DATA SINK                                                          |
|     EXCHANGE ID: 02                                                         |
|     UNPARTITIONED                                                           |
|                                                                             |
|   1:Project                                                                 |
|   |  <slot 5> : 9: abs + 1.0                                                |
|   |  <slot 6> : ceil(9: abs)                                                |
|   |  common expressions:                                                    |
|   |  <slot 7> : CAST(1: a AS DOUBLE)                                        |
|   |  <slot 8> : 7: cast + 3: pv                                             |
|   |  <slot 9> : abs(8: add)                                                 |
|   |                                                                         |
|   0:OlapScanNode                                                            |
|      TABLE: t                                                               |
|      PREAGGREGATION: ON                                                     |
|      partitions=1/1                                                         |
|      rollup: t                                                              |
|      tabletRatio=10/10                                                      |
|      tabletList=11004,11006,11008,11010,11012,11014,11016,11018,11020,11022 |
|      cardinality=1                                                          |
|      avgRowSize=4.0                                                         |
|      numNodes=0                                                             |
+-----------------------------------------------------------------------------+
34 rows in set (0.01 sec)
```

### resulsts

```
mysql> explain SELECT     sum(step0_time != '2020-01-02 00:00:00') AS step0,     sum(step1_time != '2020-01-02 00:00:00') AS step1 from( select uid, times, actions, step0_time,  array_filter((x,y)-> (y = 'buy'   AND (x BETWEEN step0_time AND date_add(step0_time, INTERVAL 90 MINUTE))  AND (step0_time <> '2020-01-01 00:0
0:00' ) ), times, actions)[1] as step1_time   from  (  select uid, times, actions,   array_filter(x -> (x = 'pv'), actions)[1] AS step0_time  from      (select
uid, array_sort(array_agg(time_stamp)) as times,  array_sortby(array_agg(action),array_agg(time_stamp)) as actions  from taobao_actions where to_date(time_stamp)='2017-11-25' group by uid) as t ) as t1) as t2;
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Explain String                                                                                                                                                                                                                                                                                                                      |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| PLAN FRAGMENT 0                                                                                                                                                                                                                                                                                                                     |
|  OUTPUT EXPRS:17: sum | 18: sum                                                                                                                                                                                                                                                                                                     |
|   PARTITION: RANDOM                                                                                                                                                                                                                                                                                                                 |
|                                                                                                                                                                                                                                                                                                                                     |
|   RESULT SINK                                                                                                                                                                                                                                                                                                                       |
|                                                                                                                                                                                                                                                                                                                                     |
|   3:AGGREGATE (update finalize)                                                                                                                                                                                                                                                                                                     |
|   |  output: sum(23: expr != '2020-01-02 00:00:00'), sum(array_filter(20: array_sort, array_map((<slot 12>, <slot 13>) -> ((<slot 13> = 'buy') AND ((<slot 12> >= 24: cast) AND (<slot 12> <= minutes_add(24: cast, 90)))) AND (23: expr != '2020-01-01 00:00:00'), 20: array_sort, 19: array_sortby))[1] != '2020-01-02 00:00:00') |
|   |  group by:                                                                                                                                                                                                                                                                                                                      |
|   |                                                                                                                                                                                                                                                                                                                                 |
|   2:Project                                                                                                                                                                                                                                                                                                                         |
|   |  <slot 19> : 19: array_sortby                                                                                                                                                                                                                                                                                                   |
|   |  <slot 20> : 20: array_sort                                                                                                                                                                                                                                                                                                     |
|   |  <slot 23> : 23: expr                                                                                                                                                                                                                                                                                                           |
|   |  <slot 24> : 24: cast                                                                                                                                                                                                                                                                                                           |
|   |  common expressions:                                                                                                                                                                                                                                                                                                            |
|   |  <slot 19> : array_sortby(7: array_agg, 6: array_agg)                                                                                                                                                                                                                                                                           |
|   |  <slot 20> : array_sort(6: array_agg)                                                                                                                                                                                                                                                                                           |
|   |  <slot 21> : array_map(<slot 10> -> <slot 10> = 'pv', 19: array_sortby)                                                                                                                                                                                                                                                         |
|   |  <slot 22> : array_filter(19: array_sortby, 21: array_map)                                                                                                                                                                                                                                                                      |
|   |  <slot 23> : 22: array_filter[1]                                                                                                                                                                                                                                                                                                |
|   |  <slot 24> : CAST(23: expr AS DATETIME)                                                                                                                                                                                                                                                                                         |
|   |                                                                                                                                                                                                                                                                                                                                 |
|   1:AGGREGATE (update finalize)                                                                                                                                                                                                                                                                                                     |
|   |  output: array_agg(5: time_stamp), array_agg(4: action)                                                                                                                                                                                                                                                                         |
|   |  group by: 1: uid                                                                                                                                                                                                                                                                                                               |
|   |                                                                                                                                                                                                                                                                                                                                 |
|   0:OlapScanNode                                                                                                                                                                                                                                                                                                                    |
|      TABLE: taobao_actions                                                                                                                                                                                                                                                                                                          |
|      PREAGGREGATION: ON                                                                                                                                                                                                                                                                                                             |
|      PREDICATES: to_date(5: time_stamp) = '2017-11-25'                                                                                                                                                                                                                                                                              |
|      partitions=0/1                                                                                                                                                                                                                                                                                                                 |
|      rollup: taobao_actions                                                                                                                                                                                                                                                                                                         |
|      tabletRatio=0/0                                                                                                                                                                                                                                                                                                                |
|      tabletList=                                                                                                                                                                                                                                                                                                                    |
|      cardinality=1                                                                                                                                                                                                                                                                                                                  |
|      avgRowSize=3.0                                                                                                                                                                                                                                                                                                                 |
|      numNodes=0                                                                                                                                                                                                                                                                                                                     |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
38 rows in set (0.05 sec)

mysql> explain SELECT uid, times, actions, step0_time, step1_time, array_filter((x, y)->(y = '支付' AND (x BETWEEN step1_time AND date_add(step1_time, INTERVAL 90 MINUTE)) AND (step1_time <> '2020-01-01 00:00:00') ), times, actions)[1] AS step2_time FROM (SELECT uid, times, actions, step0_time, array_filter( (x, y)-> (y = '下单'
AND (x BETWEEN step0_time AND date_add(step0_time, INTERVAL 90 MINUTE)) AND (step0_time <> '2020-01-01 00:00:00' ) ), times, actions)[1] AS step1_time FROM (SELECT uid, times, actions, array_filter((x, y) -> (y= '浏览' AND x BETWEEN '2020-01-02 00:00:00' AND '2020-01-02 23:59:59'), times, actions)[1] AS step0_time FROM (SELECT uid, array_sort(array_agg(time)) AS times, array_sortby(array_agg(event_type), array_agg(time)) AS actions FROM action1 GROUP BY uid) AS t ) AS t1) AS t2;
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Explain String                                                                                                                                                                                                                                                       |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| PLAN FRAGMENT 0                                                                                                                                                                                                                                                      |
|  OUTPUT EXPRS:1: uid | 6: array_sort | 7: array_sortby | 10: expr | 13: expr | 16: expr                                                                                                                                                                              |
|   PARTITION: RANDOM                                                                                                                                                                                                                                                  |
|                                                                                                                                                                                                                                                                      |
|   RESULT SINK                                                                                                                                                                                                                                                        |
|                                                                                                                                                                                                                                                                      |
|   2:Project                                                                                                                                                                                                                                                          |
|   |  <slot 1> : 1: uid                                                                                                                                                                                                                                               |
|   |  <slot 6> : 17: array_sort                                                                                                                                                                                                                                       |
|   |  <slot 7> : 18: array_sortby                                                                                                                                                                                                                                     |
|   |  <slot 10> : 21: expr                                                                                                                                                                                                                                            |
|   |  <slot 13> : 26: expr                                                                                                                                                                                                                                            |
|   |  <slot 16> : array_filter(17: array_sort, array_map((<slot 14>, <slot 15>) -> ((<slot 15> = '支付') AND ((<slot 14> >= 26: expr) AND (<slot 14> <= minutes_add(26: expr, 90)))) AND (26: expr != '2020-01-01 00:00:00'), 17: array_sort, 18: array_sortby))[1]   |
|   |  common expressions:                                                                                                                                                                                                                                             |
|   |  <slot 17> : array_sort(4: array_agg)                                                                                                                                                                                                                            |
|   |  <slot 18> : array_sortby(5: array_agg, 4: array_agg)                                                                                                                                                                                                            |
|   |  <slot 19> : array_map((<slot 8>, <slot 9>) -> (<slot 9> = '浏览') AND ((<slot 8> >= '2020-01-02 00:00:00') AND (<slot 8> <= '2020-01-02 23:59:59')), 17: array_sort, 18: array_sortby)                                                                          |
|   |  <slot 20> : array_filter(17: array_sort, 19: array_map)                                                                                                                                                                                                         |
|   |  <slot 21> : 20: array_filter[1]                                                                                                                                                                                                                                 |
|   |  <slot 22> : minutes_add(21: expr, 90)                                                                                                                                                                                                                           |
|   |  <slot 23> : 21: expr != '2020-01-01 00:00:00'                                                                                                                                                                                                                   |
|   |  <slot 24> : array_map((<slot 11>, <slot 12>) -> ((<slot 12> = '下单') AND ((<slot 11> >= 21: expr) AND (<slot 11> <= 22: minutes_add))) AND (23: expr), 17: array_sort, 18: array_sortby)                                                                       |
|   |  <slot 25> : array_filter(17: array_sort, 24: array_map)                                                                                                                                                                                                         |
|   |  <slot 26> : 25: array_filter[1]                                                                                                                                                                                                                                 |
|   |                                                                                                                                                                                                                                                                  |
|   1:AGGREGATE (update finalize)                                                                                                                                                                                                                                      |
|   |  output: array_agg(3: time), array_agg(2: event_type)                                                                                                                                                                                                            |
|   |  group by: 1: uid                                                                                                                                                                                                                                                |
|   |                                                                                                                                                                                                                                                                  |
|   0:OlapScanNode                                                                                                                                                                                                                                                     |
|      TABLE: action1                                                                                                                                                                                                                                                  |
|      PREAGGREGATION: ON                                                                                                                                                                                                                                              |
|      partitions=0/1                                                                                                                                                                                                                                                  |
|      rollup: action1                                                                                                                                                                                                                                                 |
|      tabletRatio=0/0                                                                                                                                                                                                                                                 |
|      tabletList=                                                                                                                                                                                                                                                     |
|      cardinality=1                                                                                                                                                                                                                                                   |
|      avgRowSize=3.0                                                                                                                                                                                                                                                  |
|      numNodes=0                                                                                                                                                                                                                                                      |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
39 rows in set (0.03 sec)
```
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 22 / 22 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java|22|22|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java [hunk 1]
```diff
@@ -278,72 +310,44 @@ public Integer visit(ScalarOperator scalarOperator, Void context) {
...
             return collectCommonOperatorsByDepth(1, scalarOperator);
         }
 
         // If a scalarOperator contains any non-deterministic function, it cannot be reused
         // because the non-deterministic function results returned each time are inconsistent.
-        private boolean isNonDeterministicFunction(ScalarOperator scalarOperator) {
+        // a lambda-dependent expressions also can't be reused.
+        private boolean isNonDeterministicFuncOrLambdaDependent(ScalarOperator scalarOperator,
+                                                                Set<ColumnRefOperator> lambdaArguments) {
+            if (!reuseLambda) {
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java:323]: // isLambdaDependent means whether an expression is dependent on outer lambda arguments,
        // it works when not reuse lambda arguments. For example, when reuseLambda is false, for expression
        // array_map(x -> x+1, array), the x+1 is lambda dependent, but array_map(x -> x+1, array) is not.

if we don't want to resue lambda, we can't reuse lambda-dependent expressions, like x+1, but we can reuse lambda-independent expressions, like array_map(x->x+1,array).

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java [hunk 2]
```diff
@@ -135,13 +139,26 @@ public ScalarOperator visit(ScalarOperator scalarOperator, Void context) {
...
         private ScalarOperator tryRewrite(ScalarOperator operator) {
             if (commonOperatorsMap.containsKey(operator)) {
                 return commonOperatorsMap.get(operator);
             }
             return operator;
         }
 
+        @Override
+        public ScalarOperator visitConstant(ConstantOperator literal, Void context) {
+            return tryRewrite(literal);
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java:159]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java [hunk 3]
```diff
@@ -247,18 +264,33 @@ private static class CommonSubScalarOperatorCollector extends ScalarOperatorVisi
...
+        // it works when not reuse lambda arguments. For example, when reuseLambda is false, for expression
+        // array_map(x -> x+1, array), the x+1 is lambda dependent, but array_map(x -> x+1, array) is not.
+        private boolean isLambdaDependent;
         private CommonSubScalarOperatorCollector(boolean reuseLambda) {
             this.reuseLambda = reuseLambda;
+            this.isLambdaDependent = false;
         }
 
+
         private int collectCommonOperatorsByDepth(int depth, ScalarOperator operator) {
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java:284]: introduce `hasLambdaFunction ` to contral the special logic for lambda functions.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 41 / 42 (97.62%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Projection.java|9|10|90.00%|[64]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java|28|28|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuseRule.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 41 / 42 (97.62%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Projection.java|9|10|90.00%|[64]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java|28|28|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuseRule.java|4|4|100.00%|[]|

**Review by Seaven — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java [hunk 4]
```diff
@@ -88,11 +91,15 @@ public static ScalarOperator rewriteOperatorWithCommonOperator(ScalarOperator op
      * For [a+b, a+b+c, a+b+d], the common sub operators is [a + b]
      */
     public static Map<Integer, Map<ScalarOperator, ColumnRefOperator>> collectCommonSubScalarOperators(
+            Projection projection,
             List<ScalarOperator> scalarOperators,
             ColumnRefFactory columnRefFactory, boolean reuseLambda) {
         // 1. Recursively collect common sub operators for the input operators
         CommonSubScalarOperatorCollector operatorCollector = new CommonSubScalarOperatorCollector(reuseLambda);
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java:98]: `reuseLambda` means reusing lambda-dependent sub expressions. for example, 
select array_sum(a), array_length(a) from (select array_map(x -> 2*x+1+2*x, array)as a from t), when `reuseLambda` is false, we got `a` as a common sub expr; when `reuseLambda` is true,  we got `2*x` as a lambda common sub expr.

**Review by packy92 — DISMISSED** [review]

**Review by Seaven — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by fzhedu** [comment]
run starrocks_fe_unittest

**Comment by fzhedu** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 44 / 45 (97.78%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Projection.java|9|10|90.00%|[65]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java|31|31|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuseRule.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**